src/zcash/IncrementalMerkleTree.hpp

```diff
@@ -57,25 +58,36 @@ class MerklePath {
 template<size_t Depth, typename Hash>
 class EmptyMerkleRoots {
 public:
-    EmptyMerkleRoots() {
-        empty_roots.at(0) = Hash::uncommitted();
-        for (size_t d = 1; d <= Depth; d++) {
-            empty_roots.at(d) = Hash::combine(empty_roots.at(d-1), empty_roots.at(d-1), d-1);
-        }
+    void initialize() {
+        if (!initialized) {
+     LOCK(cs_EmptyMerkleRoots_private_variables_write);
```

Comment:
This is an instance of the double-checked locking pattern, but it has no memory barriers so is undefined behaviour as it stands. It can be fixed though; see https://preshing.com/20130930/double-checked-locking-is-fixed-in-cpp11/

Reply:
It seems to me that the best approach is to just call `LOCK()` every time; I don't think it's that expensive. I don't think the complexity of what's described in that (very good!) article is justified in this case.

Reply:
@daira do you agree on just calling LOCK()? That's what I did first time, I assumed that because it was how it was done elsewhere, it would be okay here.